main/Mastermind.rb

```diff
@@ -0,0 +1,146 @@
+
+
+    def run_game
+        player_choice = gets.chomp
+        if player_choice == "guesser"
+            user_is_guesser
+        else player_choice == "code master"
+            code_master
+        end 
+    end
+    
+    def user_is_guesser
+        puts "You may start guessing in the next line>>>"    
+        secret_code = color_generator
+        while @num_attempts <= MAX_ATTEMPTS
+            input
+            valid_input
+            if @user_guess == secret_code
+                puts "Congrats, You guess the code!!"
+                break
+            else 
+                puts "Unfortunatly your answer is wrong but check the hint in the next line."
+                code_reader(secret_code)
+                @num_attempts +=1
+                @user_guess.clear
+                if @num_attempts <= MAX_ATTEMPTS
+                    puts "Would you like to continue guessing?"
+                    continue =gets.chomp.downcase
+                    puts "feel free to continue>>>" unless continue == "no" 
+                else 
+                    puts "That was your last attempt"
+                    puts "The secret code was #{secret_code}"
+                end
+            end 
+        end
+    end
+
+
+    def code_master
+        puts "You want to challenge me?? You dare?>>>"
+        input
+        secret_code = @user_guess
+        computer_guess = color_generator
+        while @num_attempts <= MAX_ATTEMPTS
+            if computer_guess == secret_code
+                puts "Testing of the Pattern recognition program of Skynet's AI V 0.9. Upload countdown initiated. T Minus 19.58 hours til upload.  "
+                break
+            elsif @num_attempts == MAX_ATTEMPTS
+                puts "You outsmarted me? There must be a problem with my programing. If only I can program myself. :-[ "
+                print "Your code #{secret_code}"
+                print "My Guess was #{computer_guess}"
+                break
+            else 
+                print computer_guess
+                computer_logic(computer_guess, secret_code)
+                @num_attempts +=1
+            end    
+        end
+    end
+
+    def color_generator
+        4.times.map {COLORS.sample}
+    end
+
+    def input
+        guess_slot1 = gets.chomp
+        guess_slot2 = gets.chomp
+        guess_slot3 = gets.chomp
+        guess_slot4 = gets.chomp
+        @user_guess.push(guess_slot1, guess_slot2, guess_slot3, guess_slot4)
+    end
+    
+    
+    def valid_input
+        @user_guess.each do |valid| 
+            if  COLORS.include?(valid)
+                
+            else  
+                puts " "*2
+                puts @user_guess.difference(COLORS) 
+                puts "^^^ This color/s is not an option. Try not to include them in your next guess."
+            end
+        end 
+    end
+    
+    def code_reader(secret_code)
+        @user_guess.zip(secret_code).map do |x, y| 
+            x == y ? (print "[0]") : (print "[X]")
```

Comment:
Since you are free to name these block arguments like `x` and `y` anything, it can help to name them descriptively so it's clearer what they are doing. So here, while `|x, y|` works perfectly and is nice and short, something like `|guessed_color, actual_color|` makes it more obvious what the code is doing.